src/mscorlib/shared/System/IO/FileStreamCompletionSource.Win32.cs

```diff
@@ -35,13 +35,33 @@ private unsafe class FileStreamCompletionSource : TaskCompletionSource<int>
             private NativeOverlapped* _overlapped; // Overlapped class responsible for operations in progress when an appdomain unload occurs
             private long _result; // Using long since this needs to be used in Interlocked APIs
 
+            private MemoryHandle _handle; // mutable struct; do not make this readonly
```

Comment:
I'd have to measure again, but just looking at the existing fields, looks like this will increase the size of the object by app 33%. For the common case of the internal buffer being used, that's unfortunate. Can we avoid that? That was one of the reasons for the separate derived type. How about just creating a factory method that contains the below logic and then use that at the two call sites rather than the current naive logic based purely on whether an array is wrapped?

Reply:
You think the one extra field is worth that complexity?

Reply:
It's not one field. It's three ptr-size fields.

Reply:
Is this created for every Read, could it be reused across Reads attached to the Stream?

Reply:
Every non-buffered read and write on an async FileStream.

Reply:
Sure, when you originally did this was it pre-emptive? What did you measure that made you split the classes? Was it just object size?

You could do something similar with the cancellation registration and remove the _cancellationRegistration field if the CanBeCancelled is false.

Reply:
> What did you measure that made you split the classes? Was it just object size?

Yes, it avoided regressing bytes allocated from previous releases.

> You could do something similar with the cancellation registration and remove the _cancellationRegistration field if the CanBeCancelled is false.

Yes, we could, but not without non-trivial extra complexity: it's accessed from the callback, not just Dispose.   Also, MemoryHandle is bigger, and from a complexity perspective it's not worth classes for all combinations. I do think it's work two classes, that we already have.

Reply:
I'll send an update

Reply:
Thanks